### PR TITLE
Add support for LLaMA-2

### LINKED ISSUE #501: Support LLaMA-2

### PR DESCRIPTION
**Pull request by zhuohan123**
Fix #501 

~Update: this PR has some correctness issues on 70B models. Will look into it.~

@WoosukKwon This PR is ready to go. Please review and let's merge it!

### CODE REVIEW

**Comment by gesanqiu** [comment]
I notice that you didn't new a llama2.py, so this PR compatible with both llama and llama2?

**Comment by zhuohan123** [comment]
> I notice that you didn't new a llama2.py, so this PR compatible with both llama and llama2?

Yes, old model should still be compatible.

File: vllm/model_executor/models/llama.py [hunk 1]
```diff
@@ -93,12 +94,20 @@ def __init__(
         assert self.total_num_heads % tensor_model_parallel_world_size == 0
         self.num_heads = (self.total_num_heads //
                           tensor_model_parallel_world_size)
+        self.total_num_key_value_heads = num_key_value_heads
+        assert (self.total_num_key_value_heads %
+                tensor_model_parallel_world_size == 0)
+        self.num_key_value_heads = (self.total_num_key_value_heads //
+                                    tensor_model_parallel_world_size)
```
Comment (WoosukKwon) [inline vllm/model_executor/models/llama.py:101]: nit: I feel the variable names are unnecessarily long and they are making the code less pretty. Can we use something like `num_kv_heads` and `tp_size` instead?

**Review by WoosukKwon — APPROVED** [review]
@zhuohan123 Awesome! Thanks for the great work! Just left minor comments.

Please double-check that this PR doesn't break LLaMA V1 and other models using RoPE, before merging the PR.

File: vllm/model_executor/models/llama.py [hunk 2]
```diff
@@ -84,6 +84,7 @@ def __init__(
         self,
         hidden_size: int,
         num_heads: int,
+        num_key_value_heads: int,
```
[the remark this answers is not in the document]
Comment (zhuohan123) [reply vllm/model_executor/models/llama.py:87]: This is to be consistent with llama's original code and the variable name in its original config.

File: csrc/pos_encoding_kernels.cu [hunk 3]
```diff
@@ -7,20 +7,21 @@ template<typename scalar_t>
 __global__ void rotary_embedding_neox_kernel(
   const int64_t* __restrict__ positions,        // [num_tokens]
   scalar_t* __restrict__ query,                 // [num_tokens, num_heads, head_size]
-  scalar_t* __restrict__ key,                   // [num_tokens, num_heads, head_size]
+  scalar_t* __restrict__ key,                   // [num_tokens, num_kv_heads, head_size]
   const scalar_t* __restrict__ cos_sin_cache,   // [max_position, 2, rot_dim // 2]
   const int rot_dim,
   const int stride,
   const int num_heads,
+  const int num_kv_heads,
```
[the remark this answers is not in the document]
Comment (zhuohan123) [reply csrc/pos_encoding_kernels.cu:15]: Just checked and they work great!

**Comment by ri938** [comment]
I am getting an error when trying to load some LLama V1 models:

LlamaConfig object has no attribute 'num_key_value_heads'

**Comment by HarrisonBT** [comment]
WARNING 07-28 03:23:18 scheduler.py:196] Input prompt (2716 tokens) is too long and exceeds limit of 4096

**Comment by tuyaao** [comment]
> I am getting an error when trying to load some LLama V1 models:
> 
> LlamaConfig object has no attribute 'num_key_value_heads'
same error from me on lastest master commit:  953f28c

**Pull request merged**